---
title: "package_setup"
author: "joshmschmidt"
date: "03/04/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
#### Setting up this package
follow steps here: https://happygitwithr.com/new-github-first.html

```{r, eval=FALSE}
library(devtools)
library(usethis)
library(roxygen2)
library(ggplot2)
library(data.table)
use_description()
use_namespace()
  usethis::use_package("roxygen2")
usethis::use_package("data.table")
usethis::use_package("Rcpp")
usethis::use_package("pacman")
usethis::use_package("parallel")
usethis::use_package("NMOF")
usethis::use_package("DescTools")
usethis::use_package("Matrix")
usethis::use_package("RcppArmadillo")
usethis::use_package("RcppEigen")
devtools::document()
usethis::use_rcpp()
usethis::use_rcpp_eigen()
usethis::use_rcpp_armadillo()
usethis::use_rcpp_eigen()

```


then add R folder to project directory, manually in R file finder

```{bash Generate msms example data, eval=FALSE}
# mu = 1e-8; rho = mu; locuslength = 20e6;Ne = 1e4; s = 0.01; additive;
# samples 30 ingroup, 1 outgroup chromosomes = 31.
msms="/Users/joshua_schmidt/msms/lib/msms.jar"
out="${HOME}/Documents/acad/acad_projects/phisweepr/inst/extdata"
java -Xmx6G -jar ${msms} 31 1 -t 8000 -r 8000 20000001 -I 2 30 1 0.01 -SAA 200 -SaA 100  -Sc 0 2 0 0 0 -SF 0 -N 10000 -seed 101010 -Smark -Sp 0.5 -oFP 0.00000000 > ${out}/msms.selection.outgroup.txt

java -Xmx6G -jar ${msms} 50 1 -t 8000 -r 8000 20000001 -SAA 1000 -SaA 500 -SF 0 0.7 -N 10000 -seed 101010 -Sp 0.5 -Smark -oFP 0.00000000 > ${out}/msms.selection.50chr.s05.b07.txt




discoal="/Users/joshua_schmidt/Documents/acad/acad_projects/selection.statistics/projectBins/discoal/discoal" 
${discoal} 31 1 1000000 -t 400 -r 400 -p 2 30 1 -ed 0.05 0 1 -ws 0.00 -a 200 -x 0.5 -c 0.7 -i 4 > ${out}/discoal.selection.outgroup.txt

discoal="/Users/joshua_schmidt/Documents/acad/acad_projects/selection.statistics/projectBins/discoal/discoal" 
${discoal} 31 1 1000000 -t 400 -r 400 -p 2 30 1 -ed 0.05 0 1 -ws 0.00 -a 200 -x 0.5 -c 0.7 -i 4 > ${out}/discoal.neutral.outgroup.txt

``` 

```{r read in and process a basic msms selection scan, eval=FALSE}
ms.data <- "/Users/joshua_schmidt/Documents/acad/acad_projects/phisweepr/inst/extdata/msms.selection.outgroup.txt"
discoal.data <- "/Users/joshua_schmidt/Documents/acad/acad_projects/phisweepr/inst/extdata/discoal.selection.outgroup.txt"
#example.data <- get.ms.output(ms.file = ms.data)
example.data <- get.simulated.data(discoal.data, "discoal")
usethis::use_data(example.data, name="simData",internal = TRUE,overwrite = TRUE)
example.data <- get.simulated.data("/Users/joshuaschmidt/Projects/phisweepr/inst/extdata/msms.selection.50chr.s1.b07.txt", "msms")
```

### NOTE *****
Check makevars file - if it has openmp, it will fuck up on mac, as clang does not support it.

```{r sfs object from genome object}
library(devtools)
library(usethis)
library(roxygen2)
library(data.table)
library(ggplot2)
library(viridis)
library(cowplot)
library(Matrix)
library(NMOF)
library(parallel)
options(scipen=999)

# genome object is example.data
# idea is to have a toplevel function 'calculate_SFSs', which returns a list object with 1-d SFS, 2-d SFS, empirical and theoretocal 
# have to estimate the mutation rate. using pi.
#load_all()
source('R/sfs_functions.R')
source('R/utilityFunctions.R')
source('R/MLconfig_functions.R')
source('R/MLfunctions.R')
Rcpp::sourceCpp('src/rcpp_sfs_functions.cpp')
Rcpp::sourceCpp('src/armaCubeFields.cpp')
Rcpp::sourceCpp('src/MLconfig_cpp_functions.cpp')
Rcpp::sourceCpp('src/ML_compute_cpp.cpp')
```


##### What are the different options we an have
estimate the SFS form the data, or assume the neutral equilibrium model.  
whether to use all sites or only polymorphic sites (and within that the use of fixed derived)  


```{r Vy + Kim generation of equilibirum SFS, fig.height= 5, fig.width= 15}
monomorphic=FALSE
fixedDerived=FALSE
load("data/example.data.rda")
dataObject <- example.data
dataObject <- append(dataObject, get_sfsDcountList(dataObject, monomorphic=monomorphic , fixedDerived = fixedDerived))
if(monomorphic==FALSE | fixedDerived==FALSE) {
  dataObject <- filterDataObject(dataObject,monomorphic=monomorphic , fixedDerived = fixedDerived)
}
dataObject[["mutation_rate"]] <- theta_from_pi(dataObject$derived_allele_counts,nSam = dataObject$sample.size, locusLength = dataObject$locusLength)
n1Min <- round(dataObject$sample.size*0.20)
n1Max <- round(dataObject$sample.size*0.80)
n1range <- n1Min:n1Max


binomial.coeffs <- CbTable(dataObject$sample.size)
subP <- subPop_freqSpec(binomial.coeffs,dataObject$sample.size)
pHomo <- calc_pHomo(binomial.coeffs,dataObject$sample.size,n1Max,n1Min)
## can get the same thing as.....
subP_2 <- Matrix::Matrix(do.call(rbind,lapply(0:dataObject$sample.size,function(x) {sfs <- numeric(dataObject$sample.size+1); if(x > 1) {sfs[1:x] <-  1/(0:(x-1))}; sfs[1] <- 0; sfs })),sparse = TRUE)
# these are very close in values, down to 2e-7 at WORST.
unlist(lapply(1:(dataObject$sample.size+1), function(x) all.equal(subP[x,],subP_2[x,],tolerance = 2e-7)))


## conditional 2d-SFS
# from observed
obs.cond.2d.sfsL <- get_two_dimensionalSFSlist(dataObject = dataObject,n1Min = 1,n1Max = dataObject$sample.size,monomorphic = monomorphic,fixedDerived = fixedDerived) # this function is SLOW
# calculate phiN
# could try rewritting as mapply or mcmapply process.
obsPhiNTable <- listMatricesToArmaCube(sfslist = obs.cond.2d.sfsL,n1Range = n1range,nSam = dataObject$sample.size)
# assumed equilibirum sfs
expPhiNTable <- standardNeutralJointSFSArmaCube(CbTable = binomial.coeffs,pHomo = pHomo, n1Min = n1Min, n1Max = n1Max, nSam = dataObject$sample.size,theta = dataObject$mutation_rate, monomorphic = monomorphic,fixedDerived = fixedDerived)

# calculate phiS
# a matrix relating derived allele n, and k i.e rows are n=0,...n=samplesize, cols are k=0...k=n1
# three ways to estimate this.
# as above, condition sites on n1 count, and then see SFS of given N. summed across all sites of n
nkMatrix <- getSfs_perN_from_two_dimensionalSFSlist(two_dimensionalSFSlist = obs.cond.2d.sfsL,monomorphic = monomorphic,fixedDerived = fixedDerived)
# downproject from the full sample size SFS
nkMatrixdp <- getSfs_perN_from_downprojectFullSFS(sfs = dataObject$sfs,scaled = TRUE,monomorphic = monomorphic,fixedDerived = fixedDerived)
# down sample from the full sample size genotype array
nkMatrixds <- getSfs_perN_from_downsampleFullSFS(dataObject = dataObject,monomorphic = monomorphic,fixedDerived = fixedDerived)

midN <- round(dataObject$sample.size/2)
plot(nkMatrix[midN+1,1:(midN+1)],pch=19,cex=0.5,ylim=c(0,0.3),ylab='pr. of sites',xlab='sf class')
points(nkMatrixdp[midN+1,1:(midN+1)],pch=19,cex=0.5,col='red')
points(nkMatrixds[midN+1,1:(midN+1)],pch=19,cex=0.5,col='seagreen')

plot(nkMatrix[dataObject$sample.size-5,1:(dataObject$sample.size-5)],pch=19,cex=0.5,ylim=c(0,0.3),ylab='pr. of sites',xlab='sf class')
points(nkMatrixdp[dataObject$sample.size-5,1:(dataObject$sample.size-5)],pch=19,cex=0.5,col='red')
points(nkMatrixds[dataObject$sample.size-5,1:(dataObject$sample.size-5)],pch=19,cex=0.5,col='seagreen')

alphaDres <- 4
log10_alphad <- make_log10alphad(minalpha=0.001,maxalpha=10,length.out = alphaDres*100+2)
phiSTable <- makePhiSTable(nSam = dataObject$sample.size,testN1s = n1range,ptable = nkMatrixds,alphad = 10^log10_alphad,beta = 1)


# define params for ML calculation
# first is resolution of alpha maximisiation
alpha_levels <- c(-6,-2,40)
pEscapeLim = 0.99999 # % pescape of a single lineage
#alphad_max_d <- maxd_byAlpha(alphaParVec=alpha_levels,pEscapeLim = pEscapeLim)

# define sites to ML
core_snps_idxs <- which(dataObject$derived_allele_counts %in% n1range)
low_alpha <- -6
# maxLRatio_table <- data.table()
# system.time(for (core_idx in core_snps_idxs){
#   ml <- ml_phir(core_idx = core_idx,
#                   dataObject = dataObject,
#                   phiNTable = obsPhiNTable,
#                   phiSTable = phiSTable,
#                   alpha_levels=alpha_levels,
#                   pEscapeLim = pEscapeLim,
#                   method="bruteforce",
#                   log10_alphad.vec = log10_alphad,
#                   low_alpha=low_alpha,
#                   monomorphic = monomorphic,
#                   minN1 = n1range[1])
#     maxLRatio_table <- rbindlist(list(maxLRatio_table,ml))
# })

system.time(mc_maxLRatio_table <- mcmapply(ml_phir,core_snps_idxs, MoreArgs = list(dataObject = dataObject,phiNTable = expPhiNTable,phiSTable = phiSTable,alpha_levels=alpha_levels,pEscapeLim = pEscapeLim,method="bruteforce",log10_alphad.vec = log10_alphad,low_alpha=low_alpha,monomorphic = monomorphic,minN1 = n1range[1]),SIMPLIFY = FALSE,mc.cores = 4))

data.table(matrix(unlist(mc_maxLRatio_table), nrow=length(mc_maxLRatio_table), byrow=T))
```

```{r }
#alphaDres c(1,2,4)
# plot escape probabilty curve
calcAlpha <- function(N,r,s) {
    alpha <- (r*log(2*N))/s
    return(log10(alpha))
}


s<-c(0.005,0.001)
col<-viridis(length(s)+1)[-3]
for(i in seq_along(s)) {
    curve(probEscape_Single_C_plot(10^calcAlpha(1e3,1e-8,s[i]),beta = 1,d=x),-0.5e6,0.5e6,n = 500, add=i!=1, col=col[i], main="Reduction of diveristy due to a selective sweep",lwd=2.5,ylab='Pr. escape a sweep',xlab='distance from sweep (bp)')
}
legend('right',legend = s, col=col, lty=1,title = 'selection coefficient',bty='n')


```

```{r convert obs cube slice to data.table and plot}
cols <- viridis(10)
# can convert a slice to a long formatDT
obsSFSlong <- matrixToLongDT(obs.cond.2d.sfsL[[16]],n1 = 15,n2 = 15)
pObs <- ggplot(obsSFSlong, aes(k1, k2)) +
  geom_tile(aes(fill = log10(pr))) +
  scale_fill_gradientn(colours = cols) + coord_equal()
```

```{r convert exp cube slice to data.table and plot}
expSFSlong <- matrixToLongDT(neutralExpSfsCube[,,8],n1 = 15,n2 = 15)

pExp <- ggplot(expSFSlong, aes(k1, k2)) +
  geom_tile(aes(fill = log10(pr))) +
  scale_fill_gradientn(colours = cols) + coord_equal()
```

```{r find residuals and plot}
# residuals
setkey(obsSFSlong,n1,k1,k2)
setkey(expSFSlong,n1,k1,k2) 
comSFSlong <- obsSFSlong[expSFSlong]
comSFSlong[,residual:=pr-i.pr]
pRes <- ggplot(comSFSlong, aes(k1, k2)) +
  geom_tile(aes(fill = residual)) +
  scale_fill_gradientn(colours = cols) + coord_equal()

plot_grid(pExp, pObs, pRes,nrow = 1,labels = c('expected', 'observed', 'residuals: obs-exp'),align = "hv")
```



```{r after selection}
neartosweep <- matrixToLongDT(phiSTable[,,4,1],11,19)
midtosweep <- matrixToLongDT(phiSTable[,,4,51],11,19)
fartosweep <- matrixToLongDT(phiSTable[,,4,102],11,19)


pNear <- ggplot(neartosweep, aes(k1, k2)) +
  geom_tile(aes(fill = log10(pr))) +
  scale_fill_gradientn(colours = cols) + coord_equal()
  
pMid <- ggplot(midtosweep, aes(k1, k2)) +
  geom_tile(aes(fill = log10(pr))) +
  scale_fill_gradientn(colours = cols) + coord_equal()
  
pFar <- ggplot(fartosweep, aes(k1, k2)) +
  geom_tile(aes(fill = log10(pr))) +
  scale_fill_gradientn(colours = cols) + coord_equal()
  pdf(file = 'effect.sweep.pdf',width = 18,height = 5)
  plot_grid(pFar, pMid, pNear,nrow = 1,labels = c('far', 'midway', 'near'),align = "hv")
  dev.off()
```



```{r expand.grid in r}
n1 <- 22
n2 = 30 - 22
k1 <- 0:n1
k2 <- 0:n2
l <- length(k1) * length(k2)
a <- array(dim=c(l,2))
for(j in seq_along(k2)){
  
}
```

